### PR TITLE
Error unmarshalling refactor

### PR DESCRIPTION
**Pull request by shorea**

### CODE REVIEW

File: protocols/aws-query-protocol/src/main/java/software/amazon/awssdk/protocols/query/unmarshall/AwsQueryErrorProtocolUnmarshaller.java [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    /**
+     * Extracts the request ID from the XML document. Request ID is a top level element
+     * for all protocols, it may be RequestId or RequestID depending on the service.
+     *
+     * @param document Root XML document.
+     * @return Request ID string or null if not present.
+     */
+    private String getRequestId(SdkHttpFullResponse response, XmlElement document) {
+        XmlElement requestId = document.getOptionalElementByName("RequestId")
+                                       .orElse(document.getElementByName("RequestId"));
```
Comment (varunnvs92) [inline protocols/aws-query-protocol/src/main/java/software/amazon/awssdk/protocols/query/unmarshall/AwsQueryErrorProtocolUnmarshaller.java]: Method in orElse is same as underlying method of getOptionalElementByName? Can we just return null in orElse()?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/test-async-client-class.java [hunk 2]
```diff
@@ -224,22 +226,23 @@ public final String serviceName() {
     public CompletableFuture<Void> eventStreamOperation(EventStreamOperationRequest eventStreamOperationRequest,
                                                         Publisher<InputEventStream> requestStream, EventStreamOperationResponseHandler asyncResponseHandler) {
         try {
+            JsonOperationMetadata operationMetadata = JsonOperationMetadata.builder().hasStreamingSuccessResponse(false)
```
Comment (varunnvs92) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/test-async-client-class.java]: Multiple instances of JsonOperationMetadata are created in this operation

File: protocols/aws-query-protocol/src/main/java/software/amazon/awssdk/protocols/query/unmarshall/AwsQueryErrorProtocolUnmarshaller.java (hunk 1 above)
Comment (shorea) [reply protocols/aws-query-protocol/src/main/java/software/amazon/awssdk/protocols/query/unmarshall/AwsQueryErrorProtocolUnmarshaller.java]: good catch, the second should have been RequestID capital D. EC2 and SimpleDB have a different casing then most XML/Query services.

File: protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/dom/SdkEmbeddedObject.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.fasterxml.jackson.core.JsonParser;
+import java.nio.ByteBuffer;
+import java.util.Date;
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * Represents an embedded object returned by a {@link JsonParser}. This is used for the ION
+ * format which embeds @{@link Date} and {@link ByteBuffer} objects.
```
[the remark this answers is not in the document]
Comment (shorea) [reply protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/dom/SdkEmbeddedObject.java]: Fixed

File: protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/JsonContent.java [hunk 4]
```diff
@@ -64,31 +65,28 @@ public static JsonContent createJsonContent(SdkHttpFullResponse httpResponse,
...
-    private static JsonNode parseJsonContent(byte[] rawJsonContent, ObjectMapper mapper) {
+    private static SdkJsonNode parseJsonContent(byte[] rawJsonContent, JsonFactory jsonFactory) {
+        JsonDomParser parser = JsonDomParser.create(jsonFactory);
         if (rawJsonContent == null || rawJsonContent.length == 0) {
-            // Note: behavior of mapper.readTree changed in 2.9 so we need to explicitly
-            // check for an empty input and return an empty object or else the return
-            // value will be null:
-            // https://github.com/FasterXML/jackson-databind/issues/1406
-            return mapper.createObjectNode();
+            // TODO andrew
```
[the remark this answers is not in the document]
Comment (shorea) [reply protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/JsonContent.java]: Actually addressed this earlier, removed todo

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/recursivestructtype.java [hunk 5]
```diff
@@ -30,7 +31,8 @@
 /**
  */
 @Generated("software.amazon.awssdk:codegen")
-public final class RecursiveStructType implements SdkPojo, ToCopyableBuilder<RecursiveStructType.Builder, RecursiveStructType> {
+public final class RecursiveStructType implements SdkPojo, Serializable,
```
[the remark this answers is not in the document]
Comment (shorea) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/recursivestructtype.java]: Possibly, I'm working on another refactor so I'll try this out when things are stable again.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/JsonProtocolSpec.java [hunk 6]
```diff
@@ -105,6 +106,12 @@ public MethodSpec initProtocolFactory(IntermediateModel model) {
         return methodSpec.build();
     }
 
+    private CodeBlock customErrorCodeFieldName() {
+        return model.getCustomizationConfig().getCustomErrorCodeFieldName() == null ?
+        CodeBlock.builder().build() :
```
[the remark this answers is not in the document]
Comment (shorea) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/JsonProtocolSpec.java]: Fixed this locally, will be included in my next commit, whatever that may be.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/runtime/transform/Marshaller.java [hunk 7]
```diff
@@ -18,8 +18,12 @@
 import software.amazon.awssdk.annotations.SdkProtectedApi;
 
 @SdkProtectedApi
-public interface Marshaller<T, R> {
+// TODO the order of these params is backwards (possibly just get rid of the marshalled type)
```
[the remark this answers is not in the document]
Comment (shorea) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/runtime/transform/Marshaller.java]: Trying to get this merged as is then I've got a few other TODOs I want to address. Right now working on getting the integ tests passing.

File: codegen/pom.xml [hunk 8]
```diff
@@ -80,11 +80,6 @@
             <groupId>org.eclipse.text</groupId>
             <version>${org.eclipse.text.version}</version>
         </dependency>
-        <dependency>
```
[the remark this answers is not in the document]
Comment (shorea) [reply codegen/pom.xml]: Right! :)

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/eventonemarshaller.java [hunk 9]
```diff
@@ -1,43 +0,0 @@
...
-import software.amazon.awssdk.core.protocol.ProtocolMarshaller;
-import software.amazon.awssdk.services.jsonprotocoltests.model.EventOne;
-import software.amazon.awssdk.utils.Validate;
-
-/**
- * {@link EventOne} Marshaller
- */
-@Generated("software.amazon.awssdk:codegen")
-@SdkInternalApi
-public class EventOneMarshaller {
```
Comment (varunnvs92) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/eventonemarshaller.java]: Should not be deleted?

File: services/s3/src/main/resources/codegen-resources/customization.config [hunk 10]
```diff
@@ -4,15 +4,9 @@
         "Object" : "S3Object"
     },
     "serviceSpecificClientConfigClass": "S3Configuration",
-    "shapeModifiers" : {
-        "GetBucketPolicyOutput" : {
-            "customUnmarshallerFqcn": "software.amazon.awssdk.services.s3.transform.GetBucketPolicyResponseUnmarshaller"
```
Comment (varunnvs92) [inline services/s3/src/main/resources/codegen-resources/customization.config]: Don't we need this?

Comment (shorea) [reply services/s3/src/main/resources/codegen-resources/customization.config]: I added an interceptor to deal with this

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/eventonemarshaller.java (hunk 9 above)
Comment (shorea) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/eventonemarshaller.java]: I just saw they were using classes that didn't exist anymore so figured they weren't used and removed them

File: services/s3/src/main/resources/codegen-resources/customization.config (hunk 10 above)
Comment (varunnvs92) [reply services/s3/src/main/resources/codegen-resources/customization.config]: Handwritten GetBucketPolicyResponseUnmarshaller class is not deleted, I got confused by that.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/eventonemarshaller.java (hunk 9 above)
Comment (varunnvs92) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/eventonemarshaller.java]: I can add them when I fix the event marshaller generation

File: services/s3/src/main/resources/codegen-resources/customization.config (hunk 10 above)
Comment (shorea) [reply services/s3/src/main/resources/codegen-resources/customization.config]: Oops, I'll delete that. Good catch.

**Pull request merged**